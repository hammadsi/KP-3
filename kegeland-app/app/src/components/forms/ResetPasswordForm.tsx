import {useForm} from 'react-hook-form';
import {Dialog, Portal, Text} from 'react-native-paper';
import React, {useCallback, useState} from 'react';
import {StyleSheet, View} from 'react-native';
import {yupResolver} from '@hookform/resolvers/yup';
import {useFocusEffect, useNavigation} from '@react-navigation/native';
import * as yup from 'yup';
import {ScrollView} from 'react-native-gesture-handler';

import useAppDispatch from '~hooks/useAppDispatch';
import {resetPassword} from '~state/ducks/auth/auth.actions';
import useAppSelector from '~hooks/useAppSelector';
import {AuthScreenProps} from '~routes/interface';
import {clearError} from '~state/ducks/auth/auth.reducer';
import Button from '~components/Button';

import FormInput from './FormInput';
import FormError from './FormError';

type ResetPasswordFormData = {
  email: string;
};

const schema = yup.object({
  email: yup.string().nullable().email().required().label('Email'),
});

/**
 * Component for rendering the reset password form.
 */
const ResetPasswordForm: React.FC = () => {
  const navigation =
    useNavigation<AuthScreenProps<'Forgot password'>['navigation']>();
  const dispatch = useAppDispatch();
  const [mailSent, setMailSent] = useState<boolean>(false);
  const {loading, error} = useAppSelector(({auth}) => auth);
  const {control, handleSubmit, formState, reset} =
    useForm<ResetPasswordFormData>({
      mode: 'onSubmit',
      resolver: yupResolver(schema),
    });

  useFocusEffect(
    useCallback(() => {
      return () => {
        setMailSent(false);
        reset();
        dispatch(clearError());
      };
    }, []),
  );

  const onSubmit = (data: ResetPasswordFormData) => {
    setMailSent(true);
    dispatch(resetPassword(data));
  };

  return (
    <ScrollView>
      <View style={styles.form}>
        <FormInput
          state={formState}
          control={control}
          testID="ResetEmailInput"
          name="email"
          placeholder="Email"
          keyboardType="email-address"
        />
      </View>
      <Button
        mode="contained"
        loading={loading}
        onPress={handleSubmit(onSubmit)}>
        Reset password
      </Button>
      <FormError error={error} />
      <Portal>
        <Dialog visible={mailSent && !loading}>
          <Dialog.Title style={styles.title}>Check your email!</Dialog.Title>
          <Dialog.Content>
            <Text textBreakStrategy="simple" style={styles.content}>
              Please check the inbox of the email address you entered for a link
              to reset your password.
            </Text>
          </Dialog.Content>
          <Dialog.Actions style={styles.actions}>
            <Button
              testID="ResetSubmit"
              onPress={() => navigation.navigate('Login')}
              labelStyle={styles.button}
              children="Login"
            />
          </Dialog.Actions>
        </Dialog>
      </Portal>
    </ScrollView>
  );
};

const styles = StyleSheet.create({
  form: {
    alignContent: 'center',
    justifyContent: 'center',
    marginVertical: 15,
  },
  title: {
    textAlign: 'center',
  },
  content: {
    textAlign: 'left',
  },
  actions: {
    justifyContent: 'center',
  },
  button: {fontSize: 18},
});

export default ResetPasswordForm;
