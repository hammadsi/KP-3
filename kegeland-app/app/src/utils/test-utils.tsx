/* istanbul ignore file */
import React from 'react';
import {NavigationContainer} from '@react-navigation/native';
import {Provider as PaperProvider} from 'react-native-paper';
import {render, RenderOptions} from '@testing-library/react-native';
import {Provider as ReduxProvider} from 'react-redux';

import {store} from '~state/store';
import {lightTheme} from '~constants/theme';
import {initialStore, mockStore} from '~state/ducks/__mocks__/store.mock';

type TestWrapperProps = {
  children: React.ReactNode;
  store?: typeof store;
};

/**
 * Wrapper component with redux state. Used when testing components
 * which requires redux.
 * @param param0 test wrapper props @see {@link TestWrapperProps}
 * @returns
 */
export const TestWrapper: React.FC<TestWrapperProps> = ({
  children,
  ...props
}) => {
  return (
    <ReduxProvider store={props.store || mockStore(initialStore)}>
      <PaperProvider theme={lightTheme}>
        <NavigationContainer theme={lightTheme}>{children}</NavigationContainer>
      </PaperProvider>
    </ReduxProvider>
  );
};

/**
 * Custom test renderer which wraps components with a redux store provider
 * @param ui the component to render
 * @param options renderer options
 * @returns test rendered componnet
 */
export const customRender = <T extends any>(
  ui: React.ReactElement<T>,
  options?: {
    wrapperProps?: {
      store: any;
    };
    renderOptions?: RenderOptions;
  },
) =>
  render(ui, {
    wrapper: (props) => <TestWrapper {...props} {...options?.wrapperProps} />,
    ...options?.renderOptions,
  });

// re-export everything
export * from '@testing-library/react-native';

// override render method
export {customRender as render};
