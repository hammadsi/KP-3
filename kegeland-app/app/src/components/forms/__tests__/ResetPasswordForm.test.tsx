import React from 'react';

import {render} from '~utils/test-utils';

import ResetPasswordForm from '../ResetPasswordForm';
jest.useFakeTimers();

describe('Test ResetPasswordForm-component', () => {
  it('should render correctly', () => {
    const component = <ResetPasswordForm />;
    const tree = render(component);
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
