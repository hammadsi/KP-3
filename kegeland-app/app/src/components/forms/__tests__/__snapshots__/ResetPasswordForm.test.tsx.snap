// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test ResetPasswordForm-component should render correctly 1`] = `
<View
  collapsable={false}
  pointerEvents="box-none"
  style={
    Object {
      "flex": 1,
    }
  }
>
  <RCTScrollView
    collapsable={false}
    handlerTag={1}
    handlerType="NativeViewGestureHandler"
    onGestureHandlerEvent={[Function]}
    onGestureHandlerStateChange={[Function]}
    waitFor={
      Array [
        Object {
          "current": null,
        },
      ]
    }
  >
    <View>
      <View
        style={
          Object {
            "alignContent": "center",
            "justifyContent": "center",
            "marginVertical": 15,
          }
        }
      >
        <View
          style={
            Object {
              "marginTop": 5,
            }
          }
        >
          <View
            style={Object {}}
          >
            <View
              pointerEvents="none"
              style={
                Array [
                  Object {
                    "bottom": 0,
                    "left": 0,
                    "position": "absolute",
                    "right": 0,
                    "top": 6,
                  },
                  Object {
                    "backgroundColor": "#f6f6f6",
                    "borderColor": "rgba(0, 0, 0, 0.54)",
                    "borderRadius": 15,
                    "borderWidth": 1,
                  },
                ]
              }
              testID="text-input-outline"
            />
            <View>
              <View
                style={
                  Array [
                    Object {
                      "paddingBottom": 0,
                    },
                    Object {
                      "minHeight": 64,
                      "paddingTop": 8,
                    },
                  ]
                }
              >
                <TextInput
                  autoCapitalize="none"
                  editable={true}
                  keyboardType="email-address"
                  maxFontSizeMultiplier={1.5}
                  multiline={false}
                  onBlur={[Function]}
                  onChangeText={[Function]}
                  onFocus={[Function]}
                  placeholder="Email"
                  placeholderTextColor="rgba(0, 0, 0, 0.54)"
                  returnKeyType="next"
                  secureTextEntry={false}
                  selectionColor="#00b0ff"
                  style={
                    Array [
                      Object {
                        "flexGrow": 1,
                        "margin": 0,
                        "paddingHorizontal": 14,
                        "zIndex": 1,
                      },
                      Object {
                        "height": 56,
                      },
                      Object {
                        "paddingBottom": 0,
                        "paddingTop": 0,
                      },
                      Object {
                        "color": "#000000",
                        "fontFamily": "System",
                        "fontSize": 16,
                        "fontVariant": undefined,
                        "fontWeight": undefined,
                        "textAlign": "left",
                        "textAlignVertical": "center",
                      },
                      false,
                      Array [
                        Object {},
                      ],
                    ]
                  }
                  testID="ResetEmailInput"
                  underlineColorAndroid="transparent"
                  value={null}
                />
              </View>
            </View>
          </View>
          <Text
            collapsable={false}
            maxFontSizeMultiplier={1.5}
            nativeID="animatedComponent"
            onLayout={[Function]}
            style={
              Object {
                "color": "#B00020",
                "fontFamily": "System",
                "fontSize": 12,
                "fontWeight": "400",
                "opacity": 0,
                "paddingHorizontal": 12,
                "paddingVertical": 4,
                "textAlign": "left",
                "transform": Array [],
                "writingDirection": "ltr",
              }
            }
          />
        </View>
      </View>
      <View
        style={
          Object {
            "backgroundColor": "#00b0ff",
            "borderColor": "transparent",
            "borderRadius": 15,
            "borderStyle": "solid",
            "borderWidth": 0,
            "elevation": 2,
            "minWidth": 64,
            "shadowColor": "#000000",
            "shadowOffset": Object {
              "height": 0.75,
              "width": 0,
            },
            "shadowOpacity": 0.24,
            "shadowRadius": 1.5,
          }
        }
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "disabled": false,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Array [
              Object {
                "overflow": "hidden",
              },
              Object {
                "borderRadius": 15,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "alignItems": "center",
                  "flexDirection": "row",
                  "justifyContent": "center",
                },
                Object {
                  "paddingVertical": 5,
                },
              ]
            }
          >
            <Text
              numberOfLines={1}
              selectable={false}
              style={
                Array [
                  Object {
                    "color": "#000000",
                    "fontFamily": "System",
                    "fontWeight": "400",
                  },
                  Object {
                    "textAlign": "left",
                  },
                  Array [
                    Object {
                      "letterSpacing": 1,
                      "marginHorizontal": 16,
                      "marginVertical": 9,
                      "textAlign": "center",
                    },
                    undefined,
                    Object {
                      "textTransform": "uppercase",
                    },
                    Object {
                      "color": "#000000",
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontFamily": "System",
                      "fontWeight": "500",
                    },
                    Object {
                      "fontSize": 16,
                    },
                  ],
                ]
              }
            >
              Reset password
            </Text>
          </View>
        </View>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
